### ISSUE TITLE
Exception type on invalid index

### ISSUE THREAD

**Issue by woutdenolf** [h5py/h5py#1119]
Invalid indexing of a dataset raises ValueError as opposed to the usual IndexError. This has unexpected consequences. Here is a MWE:
```python
import h5py

class Minimal(object):
    
    def __init__(self,dataset):
        self.dataset = dataset

    def __getitem__(self,index):
        return self.dataset[index]
            
class MinimalFix(object):
    
    def __init__(self,dataset):
        self.dataset = dataset

    def __getitem__(self,index):
        try:
            return self.dataset[index]
        except ValueError as e:
            raise IndexError(e)
        
with h5py.File('test.h5',mode='w') as f:
    f['array'] = [range(5)]*4
    for slc in MinimalFix(f['array']):
        pass
    for slc in Minimal(f['array']):
        pass
```
It's an easy fix (change ValueError with IndexError in h5py._hl_selection._translate_int). I was going to submit a PR but it breaks a lot of tests. They expect specifically ValueError exceptions. Any chance this can be changed?

**Comment by aparamon** [comment]
Your proposal does seem reasonable to me, and I'm all for consistency with Python and NumPy. But there must be consensus also with @tacaswell, @aragilar, @takluyver.

In [my recent PR](https://github.com/h5py/h5py/pull/1098), I found [h5py.get_config()](http://docs.h5py.org/en/latest/config.html#library-configuration) object really helpful. Now I think that for this and similar issues where improvement is only achievable by incompatible changes (read: string/bytes handling), `h5py.get_config()` could be (ab)used to both provide the enhancement *and* have perfect backwards compatibility by default. Do you agree that a one-liner
````py
h5py.get_config().IndexError_on_invalid_index = True
````
right after `import h5py` would be good enough compromise, in case your change might not be enabled by default?

P.S: While writing this, I realized that `h5py.get_config()` additionally providing a context manager could make it more widely usable. Not sure if it was intended this way though :-) (and threading...)

**Comment by tacaswell** [comment]
Does the exception type depend on the version of hdf5?  There is an internal mapping between the hdf5 error state and what python exception we raise.

If this changed, we should just fix it, if it has always raised `ValueError`, then we should probably instead raise

```python
class Both(ValueError, IndexError):
   ...
```

for now and think about deprecating the raising of `ValueError` in the future (it is hard to warn on though...I don't see how to warn only if the user catches the `ValueError`....).

Adding lots of global configuration is tempting, however my experience working with Matplotlib is that it can quickly get out of hand!  I do not think this is a good case for such a configuration flag.

**Comment by aragilar** [comment]
Looks like this comes from https://github.com/h5py/h5py/blob/master/h5py/_errors.pyx#L49, I think updating that line with `IndexError` should be sufficient.

**Comment by takluyver** [comment]
I guess there's no easy way to get an overview of where else that error code is used?

I think I'd just make that simple change for h5py 3.0, rather than trying to use multiple inheritance or something. I think the exception mapping in h5py is kind of a mess in any case.

**Comment by dota17** [comment]
I just investigated into [The Python Exception Class Hierarchy](https://airbrake.io/blog/python-exception-handling/class-hierarchy), They have a clear hierarchy:
+ LookupError
  + IndexError
  + KeyError
+ TypeError
+ ValueError

then I agree with @takluyver that the exception mapping in h5py is kind of a mess.

**Comment by takluyver** [comment]
This is fixed in 3.0:

```
IndexError: Index (10) out of range for (0-3)
```
